ethcore/src/ethereum/ethash.rs

```diff
@@ -180,14 +183,23 @@ impl Engine for Ethash {
 			let bound_divisor = self.ethash_params.gas_limit_bound_divisor;
 			let lower_limit = gas_limit - gas_limit / bound_divisor + 1.into();
 			let upper_limit = gas_limit + gas_limit / bound_divisor - 1.into();
-			if gas_limit < gas_floor_target {
-				min(gas_floor_target, upper_limit)
+			let gas_limit = if gas_limit < gas_floor_target {
+				let gas_limit = min(gas_floor_target, upper_limit);
+				round_block_gas_limit(gas_limit, lower_limit, upper_limit)
 			} else if gas_limit > gas_ceil_target {
-				max(gas_ceil_target, lower_limit)
+				let gas_limit = max(gas_ceil_target, lower_limit);
+				round_block_gas_limit(gas_limit, lower_limit, upper_limit)
 			} else {
-				max(gas_floor_target, min(min(gas_ceil_target, upper_limit),
-					lower_limit + (header.gas_used().clone() * 6.into() / 5.into()) / bound_divisor))
-			}
+				let total_lower_limit = max(lower_limit, gas_floor_target);
+				let total_upper_limit = min(upper_limit, gas_ceil_target);
+				let gas_limit = max(gas_floor_target, min(total_upper_limit,
```

Comment:
If I change this to `max(total_upper_limit, .. ` it will always stick to the upper_limit:
max(upper_limit, lower_limit + something) = uper_limit
So this definitely changes the logic significantly and in strange way :)
Maybe you mean `max(total_lower_limit, ...` - I see that this is logical. However:
max(N, N + M), where N && M are U256 are always larger than N (let's omit overflows)
Previously it wasn't there, because there were no `total_lower_limit` && it is an extra-op, but now everything is known && I could easily change this, if this is what you meant. Just ping me